File: 04.SERVIDOR/WS_ComElect_RESTJAVA_GR05/src/main/java/ec/edu/yaranga/comerelect/controller/ElectrodomesticoController.java
package ec.edu.yaranga.comerelect.controller;

import ec.edu.yaranga.comerelect.dto.ElectrodomesticoDTO;
import ec.edu.yaranga.comerelect.dto.ResponseDto;
import ec.edu.yaranga.comerelect.service.ElectrodomesticoService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;

import java.math.BigDecimal;
import java.util.List;

@RestController
@CrossOrigin(origins = "http://localhost:9090")
@RequestMapping("/api/electrodomesticos")
@RequiredArgsConstructor
public class ElectrodomesticoController {

    private final ElectrodomesticoService service;

    @GetMapping
    public List<ElectrodomesticoDTO> listar() {
        return service.listar();
    }

    @GetMapping("/{codigo}")
    public ElectrodomesticoDTO obtener(@PathVariable String codigo) {
        return service.obtener(codigo);
    }

    @PostMapping(consumes = {"multipart/form-data"})
    @ResponseStatus(HttpStatus.CREATED)
    public ElectrodomesticoDTO crear(
            @RequestParam("nombre") String nombre,
            @RequestParam("precio") BigDecimal precio,
            @RequestParam(value = "descripcion", required = false) String descripcion,
            @RequestParam(value = "imagen", required = false) MultipartFile imagen) {
        
        ElectrodomesticoDTO dto = new ElectrodomesticoDTO();
        dto.setNombre(nombre);
        dto.setPrecio(precio);
        dto.setDescripcion(descripcion);
        
        return service.crear(dto, imagen);
    }

    @PutMapping(value = "/{codigo}", consumes = {"multipart/form-data"})
    public ElectrodomesticoDTO actualizar(
            @PathVariable String codigo,
            @RequestParam("nombre") String nombre,
            @RequestParam("precio") BigDecimal precio,
            @RequestParam(value = "descripcion", required = false) String descripcion,
            @RequestParam(value = "imagen", required = false) MultipartFile imagen) {
        
        ElectrodomesticoDTO dto = new ElectrodomesticoDTO();
        dto.setNombre(nombre);
        dto.setPrecio(precio);
        dto.setDescripcion(descripcion);
        
        return service.actualizar(codigo, dto, imagen);
    }

    @DeleteMapping("/{codigo}")
    public ResponseEntity<ResponseDto> eliminar(@PathVariable String codigo) {
        service.eliminar(codigo);
        ResponseDto responseDto = new ResponseDto();
        responseDto.setMessage("Electrodomestico con codigo " + codigo + "eliminado exitosamente.");
        return ResponseEntity.ok(responseDto);
    }
}

File: 04.CLI MOV/CLIMOV_ComerElect_RESTJAVA_GR05/app/src/main/java/ec/edu/monster/controlador/electrodomesticos/ElectrodomesticoController.kt
package ec.edu.monster.controlador.electrodomesticos

import ec.edu.monster.modelo.ElectrodomesticoDTO
import ec.edu.monster.modelo.ResponseDto
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.http.*

interface ElectrodomesticoController {

    @GET("/api/electrodomesticos")
    suspend fun listar(): List<ElectrodomesticoDTO>

    @GET("/api/electrodomesticos/{codigo}")
    suspend fun obtener(@Path("codigo") codigo: String): ElectrodomesticoDTO

    @Multipart
    @POST("/api/electrodomesticos")
    suspend fun crear(
        @Part("nombre") nombre: RequestBody,
        @Part("precio") precio: RequestBody,
        @Part("descripcion") descripcion: RequestBody?,
        @Part imagen: MultipartBody.Part?
    ): ElectrodomesticoDTO

    @Multipart
    @PUT("/api/electrodomesticos/{codigo}")
    suspend fun actualizar(
        @Path("codigo") codigo: String,
        @Part("nombre") nombre: RequestBody,
        @Part("precio") precio: RequestBody,
        @Part("descripcion") descripcion: RequestBody?,
        @Part imagen: MultipartBody.Part?
    ): ElectrodomesticoDTO

    @DELETE("/api/electrodomesticos/{codigo}")
    suspend fun eliminar(@Path("codigo") codigo: String): ResponseDto<Any>
}
